src/components/interactive-map.js

```diff
@@ -125,153 +124,178 @@ const defaultProps = Object.assign(
+      onHover(event);
     }
 
-    if (radius) {
-      // Radius enables point features, like marker symbols, to be clicked.
-      const size = radius;
-      const bbox = [
-        [pos[0] - size, pos[1] + size],
-        [pos[0] + size, pos[1] - size]
-      ];
-      features = map && map.queryRenderedFeatures(bbox, queryParams);
-    } else {
-      features = map && map.queryRenderedFeatures(pos, queryParams);
+    const isHovering = Boolean(interactiveLayerIds && features && features.length > 0);
+    const isEntering = isHovering && !this.state.isHovering;
+    const isExiting = !isHovering && this.state.isHovering;
+
+    if (isEntering) {
+      onEvent.call(this, 'onMouseEnter', event);
+    }
+    if (isExiting) {
+      onEvent.call(this, 'onMouseLeave', event);
+    }
+    if (isEntering || isExiting) {
+      this.setState({isHovering});
     }
-    return features;
   }
+}
 
-  _onInteractionStateChange = interactionState => {
-    const {isDragging = false} = interactionState;
-    if (isDragging !== this.state.isDragging) {
-      this._updateInteractiveContext({isDragging});
-      this.setState({isDragging});
-    }
+function onPointerClick(event) {
+  const {onClick, onNativeClick, onDblClick, doubleClickZoom} = this.props;
+  let callbacks = [];
+  const isDoubleClickEnabled = onDblClick || doubleClickZoom;
+
+  // `click` is only fired on single click. `anyclick` is fired twice if double clicking.
+  // `click` has a delay period after pointer up that prevents it from firing when
+  // double clicking. `anyclick` is always fired immediately after pointer up.
+  // If double click is turned off by the user, we want to immediately fire the
+  // onClick event. Otherwise, we wait to make sure it's a single click.
+  switch (event.type) {
+    case 'anyclick':
+      callbacks.push(onNativeClick);
+      if (!isDoubleClickEnabled) {
+        callbacks.push(onClick);
+      }
+      break;
 
-    const {onInteractionStateChange} = this.props;
-    if (onInteractionStateChange) {
-      onInteractionStateChange(interactionState);
-    }
-  };
+    case 'click':
+      if (isDoubleClickEnabled) {
+        callbacks.push(onClick);
+      }
+      break;
 
-  _updateInteractiveContext(updatedContext) {
-    this._interactiveContext = Object.assign({}, this._interactiveContext, updatedContext);
+    default:
   }
 
-  _onResize = ({width, height}) => {
-    this._width = width;
-    this._height = height;
-    this._setControllerProps(this.props);
-    this.props.onResize({width, height});
+  callbacks = callbacks.filter(Boolean);
+
+  if (callbacks.length) {
+    event = normalizeEvent.call(this, event);
+    // backward compatibility: v3 `onClick` interface
+    event.features = getFeatures.call(this, event.point);
+    callbacks.forEach(cb => cb(event));
+  }
+}
+/* End of event handers */
+
+const InteractiveMap = forwardRef((props, ref) => {
+  const [controller] = useState(() => props.controller || new MapController());
```

Comment:
Maybe use `useRef` here?

Reply:
What's the difference? We just need a persistent pointer that never updates

Reply:
this looks like a simple callback, why don't you use: useCallback?

Reply:
The function passed to `useState` is the initializer.